swingset/src/main/java/com/nqadmin/swingset/utils/SSCommon.java

```diff
@@ -157,9 +166,29 @@ public void changedUpdate(final DocumentEvent de) {
 
 					removeRowSetListener();
 
+					boolean ok = true;
 					try {
-						setBoundColumnText(((javax.swing.text.JTextComponent) getSSComponent()).getText());
+						ok = setBoundColumnText(((JTextComponent) getSSComponent()).getText());
 					} finally {
+						if (!ok) {
+							// restore previous text value
+							if(previousValue != null) {
+								if (ssComponentListenerAdded) {
+									// avoid generating events while restoring text
```

Comment:
Should we manage `listenerNeedsRestoration` inline here or should we move a state tracking variable to addSSDocumentListener() and removeSSDocumentListener() similar to addSSComponentListener() and removeSSComponentListener()?

This way you can call addSSDocumentListener() and it will add the missing listener if listenerNeedsRestoration is true and then set it to false once complete.

Reply:
I don't know if that works; I'm not saying it doesn't. I'm not sure exactly what you are suggesting.

Is it possible that `addSSDocumentListener()` would be called while `listenerNeedsRestoration` is true? If it is, then I'm just confusing myself. But I don't see how that could happen.

I'm still not sure, in detail, what you're suggesting. This is more your bailiwick, LMK.

Reply:
Added a commit, https://github.com/bpangburn/swingset/pull/178/commits/eac0fc1dc82a2644121f0aa4f0238449138b4ded, with a TODO for this.